File: CppPractice/Project1/Project1/Leetcode20.cpp
#include <string>

#include <vector>

#include <iostream>

#include <stdlib.h>

#include <time.h>

#include <sstream>

#include <algorithm>

#include <map>

#include <bitset>

using namespace std;


class Solution {
public:
    bool isValid(string s) {
        vector<char> signal;
        if (s.size()==0)
        {
            return true;
        }
        for (int i = 0; i < s.size(); i++)
        {
            if (s[i]=='(')
            {
                signal.push_back(')');

            }
            else if (s[i]=='[')
            {
                signal.push_back(']');
            }
            else if (s[i] == '{')
            {
                signal.push_back('}');
            }
            else if (s[i] == ')')
            {
                if (signal.size()==0 or signal.back()==')')
                {
                    signal.pop_back();
                }
                else
                {
                    return false;
                }

            }
            else if (s[i] == ']')
            {
                if (signal.size() == 0 or signal.back() == ']')
                {
                    signal.pop_back();
                }
                else
                {
                    return false;
                }


            }
            else if (s[i] == '}')
            {
                if (signal.size() == 0 or signal.back() == '}')
                {
                    signal.pop_back();
                }
                else
                {
                    return false;
                }

            }

        }
        if (signal.size()==0)
        {
            return true;
        }

        return false;

    }
};

File: Pythonleetcode/Leetcode20.py
class Solution(object):
    def isValid(self, s):

        dict={'(':')','[':']',"{":"}"}
        find_right=[]
        for i in range(len(s)):
            if s[i] in dict:
                find_right.append(dict[s[i]])
            elif len(find_right)==0 or s[i]!=find_right[-1]:
                return False
            elif  s[i]==find_right[-1]:
                find_right.pop()


        return len(find_right)==0




print(Solution().isValid('[]{}()'))
print(Solution().isValid('[(])'))


